experimental/segmenter/src/grapheme.rs

```diff
@@ -6,23 +6,43 @@ use alloc::vec::Vec;
 use icu_provider::prelude::*;
 
 use crate::indices::{Latin1Indices, Utf16Indices};
+use crate::iterator_helpers::derive_usize_iterator;
 use crate::rule_segmenter::*;
 use crate::{provider::*, SegmenterError};
 use utf8_iter::Utf8CharIndices;
 
 /// Grapheme cluster break iterator for an `str` (a UTF-8 string).
-pub type GraphemeClusterBreakIteratorUtf8<'l, 's> = RuleBreakIterator<'l, 's, RuleBreakTypeUtf8>;
+///
+/// For more information, see [`GraphemeClusterSegmenter`].
+#[derive(Debug)]
+pub struct GraphemeClusterBreakIteratorUtf8<'l, 's>(RuleBreakIterator<'l, 's, RuleBreakTypeUtf8>);
```

Comment:
question: Why the separate types? This means that you no longer have default access to the RBI methods, which are useful for break status and such

I think separate types makes sense over FFI but for Rust it's nice to be able to abstract over different iterators and write things built on segmenters that are still encoding-agnostic

Reply:
Well there's only 1 function in common, the Iterator. Only WordBreakIterator has the rule status; the other impls GIGO. And LineBreakIterator was always an odd one out. So it seemed like an improvement to make them all separate types.

Reply:
I'm not opposed to having different types for the four different kinds of segments, I _am_ opposed to having a type for every encoding-segment combo I think.


(I also think we probably still could have one type provided it has a second parameter that selects between Word/Line/etc)

Reply:
Can we add a trait to unify them, rather than a typedef? It seems much more Rusty to do it that way.

Reply:
The trait would only do anything in the WordBreakIterator case. The other three have no functionality except a forward Iterator. When more features get added to them, I can see adding traits around it.

Reply:
@sffc unify what? I think the different _encoding_ kinds should be unified by a typedef.

I think the different _segment_ kinds can be separate types, optionally unified by a trait.

Reply:
For the different encoding kinds I think having a generic type is quite idiomatic since it also lets people plug in other things if they really want, and they can write libraries that need to use multiple break kinds in a way that are generic over encoding (for example an editor will need to run both line and grapheme segmentation, but may wish to work on files of multiple encodings).


I'm not actually convinced we need the _typedefs_, since they're iterators and you almost never need to name them, but I do think the returned type should be something like `WBI<UTF8>`

For the different segment kinds I agree, a typedef feels awkward.

Reply:
I was thinking we could have (not now but later)

```rust
pub struct SentenceBreakIteratorUtf8<'l, 's>(...);
pub struct SentenceBreakIteratorPotentiallyIllFormedUtf8<'l, 's>(...);
pub struct SentenceBreakIteratorLatin1<'l, 's>(...);
pub struct SentenceBreakIteratorUtf16<'l, 's>(...);

pub trait SentenceBreakIterator { ... }

impl<'l, 's> SentenceBreakIterator for SentenceBreakIteratorUtf8<'l, 's> { ... }
impl<'l, 's> SentenceBreakIterator for SentenceBreakIteratorPotentiallyIllFormedUtf8<'l, 's> { ... }
impl<'l, 's> SentenceBreakIterator for SentenceBreakIteratorLatin1<'l, 's> { ... }
impl<'l, 's> SentenceBreakIterator for SentenceBreakIteratorUtf16<'l, 's> { ... }
```

Can you state what your preferred outcome is?

Reply:
I think my preferred outcome is:

```rust
pub struct SentenceBreakIterator<Encoding> {..}
pub struct WordBreakIterator<Encoding> {..}
pub struct LineBreakIterator<Encoding> {..}
pub struct GraphemeBreakIterator<Encoding> {..}


// optional, unsure if necessary except at the FFI layer
pub type SBIUtf8 = SBI<Utf8>;
// ....

// optional, probably not necessary, not now
pub trait BreakIterator: Iterator<Item = usize> {}
```

Reply:
So I realized that one problem with making `Encoding` a public trait and type parameter is that it is a fairly specific trait with internal methods on it. It looks like we currently get around this via a bug in rust visibility because `RuleBreakTypeUtf8` shows up in a public signature but it isn't exported from the crate.

https://unicode-org.github.io/icu4x/docs/icu/segmenter/struct.GraphemeClusterSegmenter.html#method.segment_str

Reply:
`ZeroVecLike` is another trait that's in this category. Although that trait has a lot of methods on it, all those methods are reasonable to implement and don't themselves contain references to other internal types. In our case, the trait `RuleBreakType` (or `Encoding`) is full of references to internals of the segmenter. It is not a trait that anyone could or should implement outside of the crate, even if they knew what they were doing.

Reply:
ok in order to avoid changing things up to much, I did the following:

1. Created the four BreakIterator structs (LineBreakIterator already existed, so I just made the other three)
2. Kept them requiring the same traits
3. Kept those traits non-exported

Reply:
Okay, this works for me. We can figure out how best to expose these traits in the future.